packages/astro/src/assets/fonts/utils.ts

```diff
@@ -264,3 +277,83 @@ export function kebab(value: string) {
 export function getFamilyName(family: Pick<FontFamilyAttributes, 'name' | 'as'>): string {
 	return family.as ?? family.name;
 }
+
+/** Resolves the font family provider. If none is provided, it will infer the correct one and resolve it. */
+export async function resolveFontFamily({
+	family,
+	...resolveProviderOptions
+}: Omit<ResolveProviderOptions, 'provider'> & {
+	family: FontFamily<BuiltInProvider | FontProvider>;
+}): Promise<ResolvedFontFamily> {
+	if (family.provider === LOCAL_PROVIDER_NAME || (!family.provider && 'src' in family)) {
+		return {
+			...family,
+			provider: LOCAL_PROVIDER_NAME,
+		};
+	}
+
+	const provider =
+		family.provider === GOOGLE_PROVIDER_NAME || !family.provider
+			? await import('./providers/google.js').then((mod) => mod.google())
+			: family.provider;
+
+	return {
+		...family,
+		provider: await resolveProvider({
+			...resolveProviderOptions,
+			provider,
+		}),
+	};
+}
+
+function sortObjectByKey<T extends Record<string, any>>(unordered: T): T {
+	const ordered = Object.keys(unordered)
+		.sort()
+		.reduce((obj, key) => {
+			// @ts-expect-error
+			obj[key] = unordered[key];
+			return obj;
+		}, {} as T);
+	return ordered;
+}
+
+/**
+ * Extracts providers from families so they can be consumed by unifont.
+ * It deduplicates them based on their config and provider name, to eg.
+ * allow using the same providers with different options.
+ */
+export function familiesToUnifontProviders({
+	families,
+	hashString,
+}: {
+	families: Array<ResolvedFontFamily>;
+	hashString: (value: string) => string;
+}): Array<unifont.Provider> {
+	const map = new Map<string, unifont.Provider>();
+
+	for (const { provider } of families) {
+		if (provider === LOCAL_PROVIDER_NAME) {
+			continue;
+		}
+
+		const unifontProvider = provider.provider(provider.config);
+		const hash = hashString(
+			JSON.stringify(
+				sortObjectByKey({
+					name: unifontProvider._name,
+					...provider.config,
+				}),
+			),
+		);
+		if (map.has(hash)) {
+			continue;
+		}
+		// Makes sure every font uses the right instance of a given provider
+		// if this provider is provided several times with different options
+		unifontProvider._name += `-${hash}`;
+		provider.name = unifontProvider._name;
+		map.set(hash, unifontProvider);
+	}
+
+	return [...map.values()];
```

Comment:
Here, we using three data structures: a `Map`, an Iterable emitted by `values()`, and the final array created by the spread operator. 

I suggest a better data structure: use a `Set` for the hashes, and an array to push `unifontProvider` variables, so you'll use only two types.